Blocks: Add post-block callback arg to `traverse_and_serialize_block()`, change signature

ockham opened the pull request:
Per https://github.com/WordPress/wordpress-develop/pull/5247#discussion_r1331151853, I realized that sibling block insertion wasn't likely going to work the way I had planned. As a consequence, I had to come up with a new way to make it work.

Basically, we:
- Change the signature of the existing callback such that:
  - the function arguments are now a _reference_ to the current block (so it can be modified inline, which is important e.g. for `theme` attribute insertion), the parent block, and the previous block (instead of the block index and chunk index)
  - the return value is now a string that will be prepended to the result of the inner block traversal and serialization.
- Add a second callback argument to `traverse_and_serialize_block`, which is called _after_ the current inner block is traversed and serialized.
  - Its function arguments are a reference to the current block, the parent block, and the next block.

Its usage is demonstrated in #5261: The callback arguments are now sufficient to insert the serialized hooked blocks as strings, rather than via `insert_inner_block`, `prepend_inner block`, or `append_inner_block`.

Trac ticket: https://core.trac.wordpress.org/ticket/59412

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ockham commented:
Rough diff on top of #5247 to demo usage:

```diff
diff --git a/src/wp-includes/blocks.php b/src/wp-includes/blocks.php
index 0efa18ea7a..153888ed51 100644
--- a/src/wp-includes/blocks.php
+++ b/src/wp-includes/blocks.php
@@ -760,8 +760,10 @@ function get_hooked_blocks( $name ) {
 	return $hooked_blocks;
 }
 
-function insert_hooked_blocks( $block, $parent, $block_index, $chunk_index ) {
+// FIXME: Better name.
+function insert_hooked_blocks( &$block, $parent = null ) {
 	$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+	$markup = '';
 	foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
 		$hooked_block = array(
 			'blockName'    => $hooked_block_type,
@@ -770,17 +772,40 @@ function insert_hooked_blocks( $block, $parent, $block_index, $chunk_index ) {
 			'innerContent' => array(),
 			'innerBlocks'  => array(),
 		);
+
+		$hooked_block_markup = serialize_block( $hooked_block );
+
 		if ( 'before' === $relative_position ) {
-			insert_inner_block( $parent, $block_index, $chunk_index, $hooked_block );
-		} elseif ( 'after' === $relative_position ) {
-			insert_inner_block( $parent, $block_index + 1, $chunk_index, $hooked_block );
+			$markup .= $hooked_block_markup;
 		} elseif ( 'first_child' === $relative_position ) {
-			prepend_inner_block( $block, $hooked_block );
+			// TODO.
 		} elseif ( 'last_child' === $relative_position ) {
-			append_inner_block( $block, $hooked_block );
+			// TODO.
 		}
 	}
-	return $block;
+	return $markup;
+}
+
+// FIXME: Better name.
+function post_insert_hooked_blocks( &$block, $parent = null ) {
+	$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+	$markup = '';
+	foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+		$hooked_block = array(
+			'blockName'    => $hooked_block_type,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+
+		$hooked_block_markup = serialize_block( $hooked_block );
+
+		if ( 'after' === $relative_position ) {
+			$markup .= $hooked_block_markup;
+		}
+	}
+	return $markup;
 }
 
 /**
diff --git a/src/wp-includes/class-wp-block-patterns-registry.php b/src/wp-includes/class-wp-block-patterns-registry.php
index a83b35185b..adf2fa76bf 100644
--- a/src/wp-includes/class-wp-block-patterns-registry.php
+++ b/src/wp-includes/class-wp-block-patterns-registry.php
@@ -165,7 +165,10 @@ final class WP_Block_Patterns_Registry {
 			return null;
 		}
 
-		return $this->registered_patterns[ $pattern_name ];
+		$pattern            = $this->registered_patterns[ $pattern_name ];
+		$blocks             = parse_blocks( $pattern['content'] );
+		$pattern['content'] = traverse_and_serialize_blocks( $blocks, 'insert_hooked_blocks', 'post_insert_hooked_blocks' );
+		return $pattern;
 	}
 
 	/**
```

gziolo commented on tests/phpunit/tests/blocks/serialize.php:
```diff
@@ -94,8 +93,8 @@ public function test_traverse_and_serialize_identity_from_parsed( $original ) {
 			$blocks,
-			function ( $block ) {
-				return $block;
+			function () {
```
It's fine to omit this param now that it is optional.

ockham replied:
Yeah, but don't we have to omit the entire test then? We already have one for `traverse_and_serialize_blocks` without a callback, don't we?

gziolo replied:
This test mirrors `serialize_blocks` to ensure that they work the same way.

ockham commented:
Demonstration of how to use it over at https://github.com/WordPress/wordpress-develop/pull/5261.

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -974,28 +974,47 @@ function serialize_blocks( $blocks ) {
- * @return string String of rendered HTML.
+ * @param array    $block         A representative array of a single parsed block object. See WP_Block_Parser_Block.
+ * @param callable $pre_callback  Callback to run on each block in the tree before serialization.
+ *                                It is called with the following arguments: &$block, $parent_block.
```
I need to update this to reflect that it's now also given the `$next` block.

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -933,27 +963,53 @@ function traverse_and_serialize_block( $block, $callback ) {
+		if ( is_callable( $pre_callback ) ) {
+			$result .= call_user_func_array(
+				$pre_callback,
+				array( &$block ) // At the top level, there is no parent block to pass to the callback.
```
What about the previous or next for the top-level processing?

ockham replied:
Yeah, we might wanna add that 👍

ockham replied:
ff183b972aea34e862b65cfe5d15b9a818620e3b

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -933,27 +963,53 @@ function traverse_and_serialize_block( $block, $callback ) {
  * @since 6.4.0
  *
- * @see serialize_blocks()
+ * @see serialize_block()
```
It was correct before as the alternative is `serialize_blocks`.

ockham replied:
Ah, indeed! Good spot, I'll fix that.

ockham replied:
a21da8824c

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -933,27 +963,53 @@ function traverse_and_serialize_block( $block, $callback ) {
+ * If either callback returns a string value, it will be prepended and appended to the serialized
+ * block markup, respectively.
+ *
+ * The callbacks will receive a reference to the current block as their first argument, so that they
```
In the code there is a note that no parent element gets passed for this function. I also added a question about the next and previous as they aren't currently passed to the callback.

ockham replied:
Good point!

ockham replied:
ff183b972a

ockham replied:
Passing `$prev` and `$next` per ff183b972aea34e862b65cfe5d15b9a818620e3b. 

As for the PHPDoc, I think I'll keep it as-is, since the callbacks _will_ get the parent arg at all levels of the tree except for the top level 🙂

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -933,27 +963,53 @@ function traverse_and_serialize_block( $block, $callback ) {
+ * @param callable $post_callback Callback to run on each block in the tree after it is traversed and serialized.
+ *                                It is called with the following arguments: &$block, $parent_block, $next_block.
+ *                                Its string return value will be appended to the serialized block markup.
+ * @return string Serialized block markup.
```
Should it emphasize in the return's value that it's a list of blocks? Or we could keep the previous description about rendered HTML.

ockham replied:
> Should it emphasize in the return's value that it's a list of blocks?

Not sure what you mean by that? 😅 The return value is "just" a string (and not a list), no?

> Or we could keep the previous description about rendered HTML.

TBH I didn't like the term "rendered", since IMO, that describes a different kind of markup -- the one that we get from `render_block`. I wanted to keep the distinction between that and serialization 😬

gziolo replied:
It was mostly nitpicking that it's technically: "Serialized block or multiple blocks markup". We can go with the existing phrasing.

gziolo reviewed (approved):
This looks great. I left some nitpicks, mostly about the PHPDoc comments that are quite complex with the callbacks. I was even looking at other definitions of params that are callbacks, but I didn't find any interesting pattern to reuse here to make it nicer to present in the dev docs.

There is only one remaining question I raise, whether the top-level list of blocks, should also get the previous and the next element passed to the callbacks.

ockham committed:
Blocks: Change `traverse_and_serialize_block(s)`'s callback signature.

During work on #59399, it was discovered that ''sibling'' block insertion wasn't likely going to work the way it was planned, which required devising an alternative solution. This new solution requires some changes to `traverse_and_serialize_block(s)`:

- Change the signature of the existing callback such that:
  - the return value is a string that will be prepended to the result of the inner block traversal and serialization;
  - the function arguments are: a ''reference'' to the current block (so it can be modified inline, which is important e.g. for `theme` attribute insertion), the parent block, and the previous block (instead of the block index and chunk index).
- Add a second callback argument to `traverse_and_serialize_block(s)`, which is called ''after'' the block is traversed and serialized.
  - Its function arguments are a reference to the current block, the parent block, and the next block.

Props gziolo.
Fixes #59412. See #59313.

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/56644.

dmsnell commented on src/wp-includes/blocks.php:
```diff
@@ -886,38 +886,68 @@ function serialize_blocks( $blocks ) {
+					$post_callback,
+					array( &$inner_block, $block, $next )
+				);
+			}
```
funny enough this kind of thing is where I've been exploring the use of a generator function as a context manager.

```php
foreach ( traverse_and_serialize_block( $original_block ) as &$block ) {
	// do anything here
}
```

this kind of structure in PHP gets around some of the more awkward properties of passing objects/arrays as function arguments, or of the need to explicitly `use` closed values.

guess it probably doesn't handle the recursion that well though.
